src/xml2_node.cpp

```diff
@@ -443,22 +613,70 @@ extern "C" SEXP node_children(SEXP node_sxp, SEXP only_node_sxp) {
   END_CPP
 }
 
+int node_length_impl(SEXP x, bool only_node) {
+  NodeType type = getNodeType(x);
+
+  int out;
+
+  switch(type) {
+  case NodeType::missing:
+    out = 0;
+    break;
+  case NodeType::node: {
+    SEXP node_sxp = VECTOR_ELT(x, 0);
+    XPtrNode node(node_sxp);
+
+    out = 0;
+    for(xmlNode* cur = node->xmlChildrenNode; cur != NULL; cur = cur->next) {
+      if (only_node && cur->type != XML_ELEMENT_NODE) {
+        continue;
+      }
+      ++out;
+    }
+    break;
+  }
+  default: Rf_error("Unexpected node type");
+  }
+
+  return out;
+}
+
 // [[export]]
-extern "C" SEXP node_length(SEXP node_sxp, SEXP only_node_sxp) {
+extern "C" SEXP node_length(SEXP x, SEXP only_node_sxp) {
   BEGIN_CPP
+  NodeType type = getNodeType(x);
 
-  XPtrNode node(node_sxp);
   bool only_node = LOGICAL(only_node_sxp)[0];
 
-  int i = 0;
-  for(xmlNode* cur = node->xmlChildrenNode; cur != NULL; cur = cur->next) {
-    if (only_node && cur->type != XML_ELEMENT_NODE) {
-      continue;
+  switch(type)
+  {
+  case NodeType::missing:
+  case NodeType::node   :
+    return Rf_ScalarInteger(node_length_impl(x, only_node));
+    break;
+  case NodeType::nodeset: {
+    int n = Rf_xlength(x);
+
+    if (n == 0) {
+      return Rf_ScalarInteger(0);
+    }
```

Comment:
I think everything is fine if you remove this. Generally i try and reserve early exits for cases where they are absolutely necessary (like if your algorithm requires >=2 elements or something, like most sorting algos)

Reply:
Oh wait, I see what you are doing here. This seems odd. If the nodeset is length 0, shouldn't the output be length 0? i.e. `integer()`? In which case you could definitely just remove this then it would do the right thing.

Maybe this is some old behavior I don't know about

Reply:
I also noticed this and found it quite weird which is why I opened an issue #404. I wanted to tackle that in a separate PR, forgot to mention this here.